k2/csrc/properties.cc

```diff
@@ -78,6 +90,78 @@ bool HasSelfLoops(const Fsa &fsa) {
   return false;
 }
 
+bool CheckCycles(int32_t s, std::vector<bool> *visited,
+                 std::vector<bool> *back_arc, const std::vector<Arc> &arcs) {
+  if ((*visited)[s] == false) {
+    (*visited)[s] = true;
+    (*back_arc)[s] = true;
+
+    std::unordered_set<int32_t> adj;
+    for (const auto &arc : arcs)
+      if (arc.src_state == s)
+        adj.insert(arc.dest_state);
+
+    for (auto i = adj.begin(); i != adj.end(); ++i) {
+      if (!(*visited)[*i] && CheckCycles(*i, visited, back_arc, arcs))
+        return true;
+      else if ((*back_arc)[*i])
+        return true;
+    }
+  }
+  (*back_arc)[s] = false;
+  return false;
+}
+
+// Detect cycles using DFS traversal
+bool IsAcyclic(const Fsa &fsa, std::vector<int32_t> *order /*= nullptr*/) {
+  if (IsEmpty(fsa)) return true;
+
+  auto num_states = fsa.NumStates();
+  std::vector<int8_t> state_status(num_states, kNotVisited);
+
+  std::stack<DfsState> stack;
+  stack.push({0, fsa.arc_indexes[0], fsa.arc_indexes[1]});
+  state_status[0] = kVisiting;
+  bool is_acyclic = true;
+  while (is_acyclic && !stack.empty()) {
+    auto &current_state = stack.top();
+    if (current_state.arc_begin == current_state.arc_end) {
+      // we have finished visiting this state
+      state_status[current_state.state] = kVisited;
+      if (order != nullptr)
+        (*order).push_back(current_state.state);
+      stack.pop();
+      continue;
+    }
+    const auto &arc = fsa.arcs[current_state.arc_begin];
+    auto next_state = arc.dest_state;
+    auto status = state_status[next_state];
+    switch (status) {
+      case kNotVisited: {
+        // a new discovered node
+        state_status[next_state] = kVisiting;
+        auto arc_begin = fsa.arc_indexes[next_state];
+        stack.push({next_state, arc_begin, fsa.arc_indexes[next_state + 1]});
```

Comment:
change `arc_begin` here to `fsa.arc_indexes[next_state]`, it's confusing with below line `++current_state.arc_begin`